File: app/src/main/java/com/example/sample/sample/myapplication/PullToRefreshActivity.java
package com.example.sample.sample.myapplication;

import android.app.Activity;
import android.os.Bundle;
import android.os.Handler;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.example.sample.sample.view.MyAdapter;
import com.example.lib.effect.effect.widget.SpringListView;
import com.example.lib.effect.effect.widget.SpringRelativeLayout;
import com.example.sample.myapplication.R;


public class PullToRefreshActivity extends Activity {
    SpringRelativeLayout mSpringLayout;
    SwipeRefreshLayout mRefreshLayout;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.pull_to_refresh);

        mRefreshLayout = (SwipeRefreshLayout) findViewById(R.id.swipe_container);
        mSpringLayout = (SpringRelativeLayout) findViewById(R.id.spring_layout);
        mSpringLayout.addSpringView(R.id.listview);

        final SpringListView listView = findViewById(R.id.listview);
        //ArrayAdapter<String> adapter = new ArrayAdapter<String>(this,
        //        android.R.layout.simple_list_item_1, android.R.id.text1, values);
        MyAdapter adapter = new MyAdapter(this, values);
        listView.setAdapter(adapter);

        listView.setEdgeEffectFactory(mSpringLayout.createViewEdgeEffectFactory());
        listView.setOverScrollNested(true);
        listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
                Toast toast = Toast.makeText(PullToRefreshActivity.this, "click item " + i, Toast.LENGTH_SHORT);
                toast.show();
            }
        });
        View item = this.getLayoutInflater().inflate(R.layout.text_row_item, null, false);
        TextView headerText = item.findViewById(R.id.textView);
        listView.addHeaderView(item);
        headerText.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Toast toast = Toast.makeText(PullToRefreshActivity.this, "click header text", Toast.LENGTH_SHORT);
                toast.show();
                Log.d("PullToRefreshActivity", "header textview onClick");
            }
        });
        listView.setEdgeEffectColor(0xffffff);

        mRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                mRefreshLayout.setRefreshing(true);
                new Handler().postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        mRefreshLayout.setRefreshing(false);
                    }
                }, 300);
            }
        });

    }

    String[] values = new String[] {
            "List item 1",
            "List item 2",
            "List item 3",
            "List item 4",
            "List item 5",
            "List item 6",
            "List item 1",
            "List item 2",
            "List item 3",
            "List item 4",
            "List item 5",
            "List item 6",            "List item 1",
            "List item 2",
            "List item 3",
            "List item 4",
            "List item 5",
            "List item 6",            "List item 1",
            "List item 2",
            "List item 3",
            "List item 4",
            "List item 5",
            "List item 6",
            "List item 7"
    };

}

File: app/src/main/java/com/example/sample/sample/myapplication/PullToRefreshActivity.kt
package com.example.sample.sample.myapplication

import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.AdapterView.OnItemClickListener
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.example.lib.effect.effect.widget.SpringListView
import com.example.lib.effect.effect.widget.SpringRelativeLayout
import com.example.sample.myapplication.R
import com.example.sample.sample.view.MyAdapter
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class PullToRefreshActivity : AppCompatActivity() {
    val TAG = localClassName
    //var mSpringLayout: SpringRelativeLayout? = null
    //var mRefreshLayout: SwipeRefreshLayout? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.pull_to_refresh)
        val mRefreshLayout = findViewById<View>(R.id.swipe_container) as SwipeRefreshLayout
        val mSpringLayout = findViewById<View>(R.id.spring_layout) as SpringRelativeLayout
        mSpringLayout.addSpringView(R.id.listview)
        val listView = findViewById<SpringListView>(R.id.listview)
        //ArrayAdapter<String> adapter = new ArrayAdapter<String>(this,
        //        android.R.layout.simple_list_item_1, android.R.id.text1, values);
        //val adapter = MyAdapter(this, values)
        listView.adapter = MyAdapter(this, values)
        listView.setEdgeEffectFactory(mSpringLayout.createViewEdgeEffectFactory())
        listView.setOverScrollNested(true)
        listView.onItemClickListener = OnItemClickListener { adapterView, view, i, l ->
            val toast =
                Toast.makeText(this@PullToRefreshActivity, "click item $i", Toast.LENGTH_SHORT)
            toast.show()
        }
        val item = this.layoutInflater.inflate(android.R.layout.simple_list_item_1, null, false)
        val headerText = item.findViewById<TextView>(android.R.id.text1)
        headerText.text = "Header"
        listView.addHeaderView(item)
        headerText.setOnClickListener {
            val toast =
                Toast.makeText(this@PullToRefreshActivity, "click header text", Toast.LENGTH_SHORT)
            toast.show()
            Log.d("PullToRefreshActivity", "header textview onClick")
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            listView.setEdgeEffectColor(0xffffff)
        }
        mRefreshLayout.setOnRefreshListener {
            mRefreshLayout.isRefreshing = true

            lifecycleScope.launch {
                delay(1000)
                mRefreshLayout.isRefreshing = false
            }
            /*
            CoroutineScope(Dispatchers.Main).launch {
                delay(1000)
                mRefreshLayout.isRefreshing = false
            }
             */
            //Handler().postDelayed({ mRefreshLayout!!.isRefreshing = false }, 300)
        }
    }

    var values = arrayOf(
        "List item 1",
        "List item 2",
        "List item 3",
        "List item 4",
        "List item 5",
        "List item 6",
        "List item 1",
        "List item 2",
        "List item 3",
        "List item 4",
        "List item 5",
        "List item 6", "List item 1",
        "List item 2",
        "List item 3",
        "List item 4",
        "List item 5",
        "List item 6", "List item 1",
        "List item 2",
        "List item 3",
        "List item 4",
        "List item 5",
        "List item 6",
        "List item 7"
    )
}